### PR TITLE
registrar - additional features

### PR DESCRIPTION
**Pull request by lazedo**
- add xavp when registered returns success
- add reg_received_avp param for strict matching the contact
- add reg_contact_avp param for strict matching the contact
- add xavp to reg_fetch_contacts

modparam("usrloc", "xavp_contact", "ulattrs")
modparam("registrar", "reg_received_avp", "$avp(chk_origin)")
modparam("registrar", "reg_contact_avp", "$avp(chk_contact)")

using xavp with registered
   $avp(chk_origin) = $su;
   $avp(chk_contact) = $(ct{nameaddr.uri});
   if (registered("location","$fu")) {
       xlog("L_INFO", "$ci|log| checking $fu - $xavp(ulattrs=>account_id)");
  }

using xavp with reg_fetch_contacts
  if(reg_fetch_contacts("location", "$fu", "caller")) {
         xlog("L_INFO", "$ci|log| checking $fu - $(ulc(caller=>ulattrs=>account_id))");
   }

### CODE REVIEW

**Comment by miconda** [comment]
Some of the patches are quite big and there is little description of the purpose and benefits.  But they affect a really important module, where performances really matter. Just reading the code is not easy to guess your needs and impact on existing behavior.

Send an email to sr-dev providing more details for each of the patches.

As a recommendation, before writing a lot of code changing an important module, it is better to discuss first on sr-dev, to come up with a proper solution fitting the best of common use cases. Otherwise, you may lose time writing code that will not be accepted because it affects performances or the existing behavior that is commonly needed.

**Comment by miconda** [comment]
You misunderstood my proposal. I wanted a new parameter to the function: registered(talble, uri, flags), because I want to avoid searching through xavps if not needed. The parameter is read only if provided. 

If I got right the patches, you search for the flags in xavp list.

**Comment by lazedo** [comment]
i don't think i did, the flags are in params.

i added the extra search for "override" to give more flexibility.
i can put that flexibility param dependent.

**Comment by miconda** [comment]
I don't see any new parameter to registered(...) function. Do I miss something?

Perhaps I can change (sometime during next days) the current call-id matching to using a new flags param to registered() function and xavp. Then you can add from there what you need extra.

**Comment by lazedo** [comment]
there are two params
match_flags as you proposed
1 - callid
2 - received
3 - contact

the registered() function check that flag, checks the values in xavp_cfg
and assigns the value to a temp vars, that are then compared against the
value in the record.

match_return_flags controls if the record xavp is added to the transaction.

**Comment by lazedo** [comment]
i added a 3rd param to control if we search for "override" values for flags.
default is 0, don't search

**Comment by miconda** [comment]
We are not on the same meaning of terms here. I refer to function parameter, not module parameter. I want to do:
```
registered("location", "$fu", "1")
```
in order to do registered() with call-id match. The new parameter can be int or pv (see fixup_igp_null() and fixup_get_ivalue() helper functions)

**Comment by lazedo** [comment]
@miconda i believe this is it ?

**Comment by lazedo** [comment]
@miconda can you check if this is acceptable ? thanks.

**Comment by miconda** [comment]
A bit hard to follow, because now here are patches to patches to patches. Would be good to have a single patch against the master in kamailio for review.

One thing I noticed, that some flags are still a module parameter. Wouldn't it be more flexible to be part of the function parameter, different flag values (eventually starting at 8 or 16)?

**Comment by lazedo** [comment]
@miconda hi, the remaining flag in the module is to control the setting of the contact xavp on return from a match. i left it here because we also want the behavior to apply to reg_fetch_contacts, is that ok ?
i can make a simple patch against master.

**Comment by miconda** [comment]
But reg_fetch_contacts() is already returning the contacts in an variable structure - $ulc. I don't see a reason to add to $ulc as well as to an xavp.

Could be better to have a new function reg_fetch_contact_to_xavp() that will get the attributes in xavp instead of $ulc.

I may miss some point, in that case, provide some details when you find useful for reg_fetch_contacts() to return some data twice.

**Comment by lazedo** [comment]
@miconda the $ulc structure doesn't have the contact xavp (it copies all elements but the xavp), the purpose of the flag is to add it to the $ulc structure.

**Comment by miconda** [comment]
But that was a patch adding to $ulc(...), not to an $xavp(), right? Maybe seeing all the changes as a single patch will shed light on these aspects.

**Comment by lazedo** [comment]
right, it doesn't add it to xavp it adds it to $ulc so if we have
modparam("usrloc", "xavp_contact", "ulattrs")
we can use the $qlc returned like this (if flag is set).
if(!reg_fetch_contacts("location", "$fu", "caller")) {
      xlog("L_INFO", "$ci|log| checking $fu - $ulc(caller=>ulattrs=>account_id)");
}

i'll put everything in one commit and create a new PR

**Comment by lazedo** [comment]
https://github.com/kamailio/kamailio/pull/101
